staking/entryWriter.go

```diff
@@ -0,0 +1,196 @@
+package staking
+
+import (
+	"errors"
+	"fmt"
+	"github.com/pegnet/pegnet/spr"
+	log "github.com/sirupsen/logrus"
+	"sync"
+
+	"github.com/FactomProject/factom"
+	"github.com/cenkalti/backoff"
+	"github.com/pegnet/pegnet/common"
+	"github.com/zpatrick/go-config"
+)
+
+type IEntryWriter interface {
+	PopulateECAddress() error
+	NextBlockWriter() IEntryWriter
+	SetSPR(spr *spr.StakingPriceRecord)
+	CollectAndWrite(blocking bool)
+	ECBalance() (int64, error)
+}
+
+// EntryWriter writes the SPRs to factom once the staking is done
+type EntryWriter struct {
+	// We need an spr template to make the entries
+	sprTemplate *spr.StakingPriceRecord
+
+	ec     *factom.ECAddress
+	config *config.Config
+
+	Next *EntryWriter
+
+	EntryWritingFunction func() error
+
+	sync.Mutex
+	sync.Once
+}
+
+func NewEntryWriter(config *config.Config) *EntryWriter {
+	w := new(EntryWriter)
+	w.config = config
+	w.EntryWritingFunction = w.writeStakingRecord
+	return w
+}
+
+func (w *EntryWriter) ECBalance() (int64, error) {
+	return factom.GetECBalance(w.ec.String())
+}
+
+// PopulateECAddress only needs to be called once
+func (w *EntryWriter) PopulateECAddress() error {
+	// Get the Entry Credit Address that we need to write our SPR records.
+	if ecadrStr, err := w.config.String("Miner.ECAddress"); err != nil {
+		return err
+	} else {
+		ecAdr, err := factom.FetchECAddress(ecadrStr)
+		if err != nil {
+			return err
+		}
+		w.ec = ecAdr
+	}
+	return nil
+}
+
+// NextBlockWriter gets the next block writer to use for the staker.
+func (w *EntryWriter) NextBlockWriter() IEntryWriter {
+	w.Lock()
+	defer w.Unlock()
+	if w.Next == nil {
+		w.Next = NewEntryWriter(w.config)
+		w.Next.ec = w.ec
+	}
+	return w.Next
+}
+
+// SetSPR is here because we need an spr to create the entry.
+func (w *EntryWriter) SetSPR(spr *spr.StakingPriceRecord) {
+	w.Lock()
+	defer w.Unlock()
+	if w.sprTemplate == nil {
+		w.sprTemplate = spr.CloneEntryData()
+	}
+}
+
+// CollectAndWrite will write the block when we collected all the staker data
+//	The blocking is mainly for unit tests.
+func (w *EntryWriter) CollectAndWrite(blocking bool) {
+	w.Do(func() {
+		if blocking {
+			w.collectAndWrite()
+		} else {
+			go w.collectAndWrite()
+		}
+	})
+}
+
+// collectAndWrite is idempotent
```

Comment:
It has side-effect by writing to blockchain,  `idempotent` incorrect wording here. Please add more broad comment on what it does